xcode/Subconscious/Shared/Services/AddressBookService.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  AddressBookService.swift
+//  Subconscious
+//
+//  Created by Ben Follington on 14/3/2023.
+//
+
+import Foundation
+import Combine
+// temp
+import SwiftUI
+
+class AddressBookService {
+    private(set) var noosphere: NoosphereService
+    private var addressBook: [AddressBookEntry]?
+    
+    init(noosphere: NoosphereService) {
+        self.noosphere = noosphere
+        self.addressBook = nil
+    }
+    
+    /// Get the full list of entries in the address book.
+    /// This is cached after the first use unless requested using refetch.
+    /// If an error occurs producing the entries the resulting list will be empty.
+    func listEntries(refetch: Bool = false) -> AnyPublisher<[AddressBookEntry], Never> {
+        if let entries = addressBook {
+            if !refetch {
+                return Just(entries).eraseToAnyPublisher()
+            }
+        }
+        
+        return self.listPetnamesAsync()
+            .flatMap { follows in
+                return Publishers.MergeMany(follows.map { f in
+                    return self.getPetnameAsync(petname: f)
+                        .map { did -> AddressBookEntry? in
+                            guard let did = did else { return nil }
+                            // TODO: hardcoded dog-pfp
+                            return AddressBookEntry(pfp: Image("dog-pfp"), petname: f, did: did)
+                        }
+                        .compactMap { $0 }
```

Comment:
Comment: I typically write `.compactMap({ value in value })` just because it is more familiar to people who might not know the Swift sugars.

No change necessary.